route-rs-runtime/src/link/classify_link.rs

```diff
@@ -1,64 +1,140 @@
 use crate::element::Classifier;
 use crate::link::task_park::*;
-use crate::link::{QueueEgressor, PacketStream};
+use crate::link::{Link, LinkBuilder, PacketStream, QueueEgressor};
 use crossbeam::atomic::AtomicCell;
 use crossbeam::crossbeam_channel;
 use crossbeam::crossbeam_channel::{Receiver, Sender};
 use futures::{Async, Future, Poll, Stream};
 use std::sync::Arc;
 
-pub struct ClassifyLink<'a, C: Classifier> {
-    pub ingressor: ClassifyIngressor<'a, C>,
-    pub egressors: Vec<QueueEgressor<C::Packet>>,
+pub struct ClassifyLink<C: Classifier> {
+    in_stream: Option<PacketStream<C::Packet>>,
+    classifier: Option<C>,
+    dispatcher: Option<Box<dyn Fn(C::Class) -> usize + Send + Sync + 'static>>,
+    queue_capacity: usize,
+    branches: Option<usize>,
 }
 
-impl<'a, C: Classifier> ClassifyLink<'a, C> {
-    pub fn new(
-        input_stream: PacketStream<C::Packet>,
-        classifier: C,
-        dispatcher: Box<dyn Fn(C::Class) -> usize + Send + Sync + 'a>,
-        queue_capacity: usize,
-        branches: usize,
+impl<C: Classifier> ClassifyLink<C> {
+    pub fn new() -> Self {
+        ClassifyLink {
+            in_stream: None,
+            classifier: None,
+            dispatcher: None,
+            queue_capacity: 10,
+            branches: None,
+        }
+    }
+
+    pub fn classifier(self, classifier: C) -> Self {
+        ClassifyLink {
+            in_stream: self.in_stream,
+            classifier: Some(classifier),
+            dispatcher: self.dispatcher,
+            queue_capacity: self.queue_capacity,
+            branches: self.branches,
+        }
+    }
+
+    pub fn dispatcher(
+        self,
+        dispatcher: Box<dyn Fn(C::Class) -> usize + Send + Sync + 'static>,
     ) -> Self {
+        ClassifyLink {
+            in_stream: self.in_stream,
+            classifier: self.classifier,
+            dispatcher: Some(dispatcher),
+            queue_capacity: self.queue_capacity,
+            branches: self.branches,
+        }
+    }
+
+    pub fn queue_capacity(self, queue_capacity: usize) -> Self {
         assert!(
-            branches <= 1000,
-            format!("ClassifyLink branches: {} > 1000", branches)
+            (1..=1000).contains(&queue_capacity),
+            format!(
+                "Queue capacity: {}, must be in range 1..=1000",
+                queue_capacity
+            )
         );
+        ClassifyLink {
+            in_stream: self.in_stream,
+            classifier: self.classifier,
+            dispatcher: self.dispatcher,
+            queue_capacity,
+            branches: self.branches,
+        }
+    }
+
+    pub fn branches(self, branches: usize) -> Self {
         assert!(
-            queue_capacity <= 1000,
-            format!("ClassifyLink queue_capacity: {} > 1000", queue_capacity)
+            (2..=1000).contains(&branches),
+            format!("Branches: {}, must be in range 2..=1000", branches)
```

Comment:
"must be between 2 and 1000 branches"?

Reply:
At first I thought I would protect the user from making a 1-1 classify cause what is the point? Same for 1-1 CloneLink. Decided to remove that cause it isn't completely the libraries job to protect the user from stupid inefficiency. There may be a use we aren't considering.